scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,24 +1089,74 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
+      mergeAvoidingHashes(merged)
+    }
+  }
+
+  /**
+   * Special handling for cases where one side of the hashjoin is merged
+   *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+   *  so we insert checkpoints and/or intermediate merge stages as appropriate
+   *
+   * @param pipes a list of pipes possibly colliding with each other
+   * @return an updated list of pipes, which can safely be merged
+   */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
```

Comment:
can we use private rather than protected? Also, these are all really static methods, right? Can we move them to an object that is `private[scalding]`? I'd rather not add them to `TypedPipe` since they seem to be working just on `Pipe`. Maybe `RichPipe`'s object?